### PR TITLE
Add x509_crl_info module

### PR DESCRIPTION
**Pull request by felixfontein**
##### SUMMARY
Follow-up to #63435. This adds a basic x509_crl_info module, and moves some common code to module_utils/crypto.py, in particular `get_relative_time_option()` (which exists in slightly different versions in several modules).

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
x509_crl_info

### CODE REVIEW

**Comment by felixfontein** [comment]
I'm currently wondering whether a query functionality (is this certificate revoked?) should be part of this module (similar to valid_at in openssl_certificate_info), or whether there should be an additional x509_crl_query module. Any opinions?

**Comment by felixfontein** [comment]
I created a new PR for the refactoring part: #67540

**Comment by ctrufan** [comment]
You wouldn't want a revocation check to be CRL specific, but support OCSP as well. I'd lean towards a separate certificate_revocation_info module (or put it in *certificate_info as an optional check).

**Comment by felixfontein** [comment]
@ctrufan good point! I think it's probably better to have a separate `openssl_certificate_revocation_info` module, since the input can also be different than for `openssl_certificate_info` - like only having the serial number of the certificate, something that won't work for `openssl_certificate_info`.

File: lib/ansible/modules/crypto/x509_crl_info.py [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+
+        self.module = module
+
+        self.crl = None
+        if self.content is None:
+            try:
+                with open(self.path, 'rb') as f:
+                    data = f.read()
+            except Exception as e:
+                self.module.fail_json(msg='Error while reading CRL file from disk: {0}'.format(e))
```
Comment (Andersson007) [inline lib/ansible/modules/crypto/x509_crl_info.py:181]:
```suggestion
                self.module.fail_json(msg='Error while reading CRL file from disk: {0}'.format(to_native(e)))
```

File: lib/ansible/modules/crypto/x509_crl_info.py [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+
+TIMESTAMP_FORMAT = "%Y%m%d%H%M%SZ"
+
+
+class CRLError(crypto_utils.OpenSSLObjectError):
+    pass
+
+
+class CRLInfo(crypto_utils.OpenSSLObject):
+
```
Comment (Andersson007) [inline lib/ansible/modules/crypto/x509_crl_info.py:162]:
```suggestion
    """Following the Ansible dev guide, better to put a docstring here."""
```

File: lib/ansible/modules/crypto/x509_crl_info.py [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+        else:
+            data = self.content.encode('utf-8')
+
+        try:
+            self.crl = x509.load_pem_x509_crl(data, default_backend())
+        except Exception as e:
+            self.module.fail_json(msg='Error while decoding CRL: {0}'.format(e))
+
+    def _dump_revoked(self, entry):
+        return {
```
Comment (Andersson007) [inline lib/ansible/modules/crypto/x509_crl_info.py:191]: i don't know why but when i'm looking at ```_dump_revoked()```, my brain wants to scroll it down as fast as possible:)
maybe would be better to split it somehow
not critical

File: lib/ansible/modules/crypto/x509_crl_info.py [hunk 4]
```diff
@@ -0,0 +1,280 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+
+# Copyright: (c) 2019, Felix Fontein <felix@fontein.de>
```
Comment (Andersson007) [inline lib/ansible/modules/crypto/x509_crl_info.py]:
```suggestion
# Copyright: (c) 2020, Felix Fontein <felix@fontein.de>
```

File: lib/ansible/modules/crypto/x509_crl_info.py (hunk 2 above)
Comment (Andersson007) [reply lib/ansible/modules/crypto/x509_crl_info.py:162]: even short doc strings for the methods would be also good

File: lib/ansible/modules/crypto/x509_crl_info.py (hunk 1 above)
Comment (Andersson007) [reply lib/ansible/modules/crypto/x509_crl_info.py:181]: and in a few places below

**Review by Andersson007 — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/crypto/x509_crl_info.py [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+    if not CRYPTOGRAPHY_FOUND:
+        module.fail_json(msg=missing_required_lib('cryptography >= {0}'.format(MINIMAL_CRYPTOGRAPHY_VERSION)),
+                         exception=CRYPTOGRAPHY_IMP_ERR)
+
+    try:
+        crl = CRLInfo(module)
+        result = crl.get_info()
+        module.exit_json(**result)
+    except crypto_utils.OpenSSLObjectError as exc:
+        module.fail_json(msg=to_native(exc))
```
Comment (Andersson007) [inline lib/ansible/modules/crypto/x509_crl_info.py]:
```suggestion
    except crypto_utils.OpenSSLObjectError as e:
        module.fail_json(msg=to_native(e))
```
imo better to use one style for similar things if not necessary to use different

File: lib/ansible/modules/crypto/x509_crl_info.py (hunk 2 above)
Comment (felixfontein) [reply lib/ansible/modules/crypto/x509_crl_info.py:162]:
```suggestion
    """The main module implementation."""

```

File: lib/ansible/modules/crypto/x509_crl_info.py (hunk 1 above)
Comment (felixfontein) [reply lib/ansible/modules/crypto/x509_crl_info.py:181]: Why is that necessary? The code works fine both in Python 2 and Python 3 (see all the other crypto modules).

File: lib/ansible/modules/crypto/x509_crl_info.py (hunk 3 above)
Comment (felixfontein) [reply lib/ansible/modules/crypto/x509_crl_info.py:191]: I'll do more refactoring once the `community.crypto` collection exists and stuff is in there. It doesn't make sense to start making `x509_crl_info` totally different from `x509_crl` right now when it is much easier doing that in the collection (by adding a new `module_utils`). :)

**Review by Andersson007 — APPROVED** [review]
shipit

File: lib/ansible/modules/crypto/x509_crl_info.py (hunk 3 above)
Comment (Andersson007) [reply lib/ansible/modules/crypto/x509_crl_info.py:191]: sounds reasonable

File: lib/ansible/modules/crypto/x509_crl_info.py (hunk 1 above)
Comment (Andersson007) [reply lib/ansible/modules/crypto/x509_crl_info.py:181]: it's not necessary but you use it in the module in other places:)

Comment (felixfontein) [reply lib/ansible/modules/crypto/x509_crl_info.py:181]: Ah, I didn't notice :) The module is mostly a copy'n'paste of `x509_crl` with many parts removed, and that one is partially copied from other `openssl_*` modules. I guess it's time to do a big cleanup everywhere. But that should also wait until after the move ;)

**Comment by MarkusTeufelberger** [comment]
Since @Andersson007 's vote probably won't count for the bot, here's my

shipit

**Comment by felixfontein** [comment]
@ctrufan @Andersson007 @MarkusTeufelberger thanks a lot for commenting and reviewing!

I'll start cleaning this more up once this has moved to the collection repo...

**Pull request merged**